### PR TITLE
remove interval tree

### PR DESCRIPTION
**Pull request by yjp20**
No more interval tree, performance is better

before w/ caching:

```
parse: 3.9278101921081543
bb build: 1.0947425365447998
bb query: 5.380667686462402
```

after w/o caching:

```
parse: 2.684222936630249 // pares was crazy fast here, I have no idea why
bb build: 0.9135973453521729
bb query: 0.5692892074584961
```

after w caching:

```
parse: 4.0205397605896
bb build: 1.0323891639709473
bb query: 0.5581090450286865
```

not much better. I did previously try to add a `@functools.lru_cache` onto the `BasicBlock` dataclass, but it didn't go super well for me during testing due to hashing issues. My benchmark ran really fast, `bb query: 0.20`, but I think I did something wrong rather than that actually being super fast.

### CODE REVIEW

File: tidalsim/bb/elf.py [hunk 1]
```diff
@@ -104,42 +103,29 @@ def get_target_from_control_instr(match: Match) -> Tuple[Optional[int], Optional
...
-
-    def end_block_after(control_instr: ObjdumpInstrEntry) -> None:
-        all_basic_blocks.slice(get_next_pc(control_instr), get_split_bbid)
+def do_basic_block_analysis(all_control_instrs: List[ObjdumpInstrEntry], initial_intervals: List[Interval]) -> List[Marker]:
+    events = intervals_to_events(initial_intervals)
 
     for control_instr in tqdm(all_control_instrs):
         if control_instr.target is None:
-            end_block_after(control_instr)
+            events += [cast(Event, (control_instr.pc, None))]
```
Comment (vighneshiyer) [inline tidalsim/bb/elf.py:119]: Can we avoid using cast if we annotate the type of `events`?

Comment (yjp20) [reply tidalsim/bb/elf.py:119]: Not a pythonista though, so take with a grain of salt 🧂.

File: tidalsim/bb/spike.py [hunk 2]
```diff
@@ -41,50 +42,35 @@ def spike_trace_to_bbs(trace: Iterator[SpikeTraceEntry]) -> BasicBlocks:
...
+    return BasicBlocks(markers=intervals_to_markers(intervals))
 
 def spike_trace_to_embedding_df(trace: Iterator[SpikeTraceEntry], bb: BasicBlocks, interval_length: int) -> DataFrame[EmbeddingSchema]:
     # Dimensions of dataframe
     # # rows = # of intervals = ceil( (length of trace) / interval_length )
     # # cols = # of features = # of elements in the intervaltree
-    n_features = len(bb.pc_to_bb_id)
-
-    # Use a cache to avoid querying the interval tree too often, queries should have good locality
-    @functools.lru_cache(maxsize=128)
```
Comment (vighneshiyer) [inline tidalsim/bb/spike.py:79]: Could you leave this cache in place? In my experiment, I see a throughput of 480 intervals/second without the cache and ~700 intervals/second with the cache. So this seems marginally faster than the intervaltree solution, but it is much cleaner imo and avoids a dependency :)

File: tidalsim/bb/spike.py [hunk 3]
```diff
@@ -1,18 +1,19 @@
...
 from more_itertools import ichunked
-import pandas as pd
 from pandera.typing import DataFrame
-from sklearn.preprocessing import normalize
 
-from tidalsim.bb.common import BasicBlocks, control_insts
+from tidalsim.bb.common import BasicBlocks, control_insts, intervals_to_markers
 from tidalsim.modeling.schemas import *
 
+# Regex patterns to extract instructions or symbols from Spike dump
```
Comment (vighneshiyer) [inline tidalsim/bb/spike.py:13]: You can get rid of these regexes since they are unused

File: tidalsim/bb/common.py [hunk 4]
```diff
@@ -1,9 +1,72 @@
...
+    return events
+
+def events_to_markers(events: List[Event]) -> List[Marker]:
+    left = 0
+    right = 0
+    idx = 0
+    markers = []
+
+    # We sort events from left-to-right, but only really care about the highest value at each location
+    for pc, valid in sorted(events, key=lambda tup: (tup[0], -tup[1])):
```
Comment (vighneshiyer) [inline tidalsim/bb/common.py:47]: Something is broken here when using `--elf`.

```
  File "/scratch/vighneshiyer/chipyard/tools/tidalsim/tidalsim/bb/common.py", line 47, in <lambda>
    for pc, valid in sorted(events, key=lambda tup: (tup[0], -tup[1])):
TypeError: bad operand type for unary -: 'NoneType'
```

Could you test this by cloning Chipyard (https://github.com/ucb-bar/chipyard) (follow instructions here: https://github.com/euphoric-hardware/tidalsim#setup-in-chipyard) and building the embench benchmarks `cd software/embench && ./build.sh`, and then invoke tidalsim like `tidalsim --elf --binary ../../software/embench/build/aha-mont64 --interval-length 1000 --clusters 18 --simulator ../../sims/vcs/simv-inject-chipyard.harness-RocketConfig --chipyard-root ../../. --dest-dir runs`.

**Review by vighneshiyer — CHANGES_REQUESTED** [review]
Thanks for working on this! The implementation is clean and the comments are helpful. The tests also look good to me. There is one issue when doing elf-based BB extraction, for which you should try to replicate and fix. I've confirmed that the basic blocks identified with this algorithm lead to good clustering results and produce similar estimated IPC traces as before.

File: tidalsim/bb/elf.py (hunk 1 above)
Comment (vighneshiyer) [reply tidalsim/bb/elf.py:119]: Yeah I think type aliases are a bit poorly handled, this is all good

File: tidalsim/bb/spike.py (hunk 2 above)
Comment (yjp20) [reply tidalsim/bb/spike.py:79]: that seems very slow to me 🤔. What were the original benchmarks and what does intervals/second mean here?

Comment (vighneshiyer) [reply tidalsim/bb/spike.py:79]: The benchmark used is the aha-mont64 binary from embench. The interval length is 1000 instructions and intervals/second means how many such intervals we were able to compute the embedding for each second by processing the spike commit log. So it basically means the current implementation can process ~700k RISC-V instructions / second when performing embedding.

**Review by vighneshiyer — APPROVED** [review]
Everything looks good! I verified that the results are similar and it appears that even a small bug was fixed.

Before:
wikisort1

After:
wikisort2

**Pull request merged**